source/proxy_connection.c

```diff
@@ -1204,6 +1204,25 @@ static int s_connect_proxy(const struct aws_http_client_connection_options *opti
         return AWS_OP_ERR;
     }
 
+    if (options->proxy_options->no_proxy_hosts.len > 0) {
+        struct aws_string *no_proxy_host_str = aws_string_new_from_cursor(options->allocator, &options->proxy_options->no_proxy_hosts);
+        if (aws_http_host_matches_no_proxy(options->allocator, options->host_name, no_proxy_host_str)) {
+            AWS_LOGF_DEBUG(
+                AWS_LS_HTTP_CONNECTION,
+                "Host \"" PRInSTR "\" found in NO_PROXY, bypassing proxy",
+                AWS_BYTE_CURSOR_PRI(options->host_name));
+            aws_string_destroy(no_proxy_host_str);
+
+            // host matched no_proxy, connect without a proxy.
+            // TODO: We need this to be null to connect without proxy, but is this correct?
+            /* Fill in a new connection options with NULL proxy_options */
+            struct aws_http_client_connection_options options_copy = *options;
+            options_copy.proxy_options = NULL; 
+            return aws_http_client_connect_internal(&options_copy, NULL);
+        }
+        aws_string_destroy(no_proxy_host_str);
```

Comment:
I don't think we need to override the proxy options here, as we won't check for proxy options later.